task_processing/plugins/kubernetes/utils.py

```diff
@@ -63,7 +66,48 @@ def is_secret_env_var(value: str) -> bool:
     return re.match(SECRET_VALUE_REGEX, value) is not None
 
 
-def get_kubernetes_env_vars(environment: PMap[str, str]) -> List[V1EnvVar]:
+def get_secret_name_from_ref(value: str) -> str:
+    return value.split("(")[1][:-1]
+
+
+def is_shared_secret(value: str) -> bool:
+    return value.startswith("SHARED_")
+
+
+def get_secret_kubernetes_env_var(
+    key: str, value: str, task_prefix: str, namespace: str,
+) -> V1EnvVar:
+    task_prefix = get_sanitised_kubernetes_name(task_prefix)
+    secret = get_secret_name_from_ref(value)
+    sanitised_secret = get_sanitised_kubernetes_name(secret)
+    if is_shared_secret(value):
+        task_prefix = get_sanitised_kubernetes_name(SHARED_SECRET_SERVICE)
+        return V1EnvVar(
+            name=key,
+            value_from=V1EnvVarSource(
+                secret_key_ref=V1SecretKeySelector(
+                    name=f"{namespace}-secret-{task_prefix}-{sanitised_secret}",
```

Comment:
is this needed?

Reply:
something like: 
```python
    task_prefix = task_prefix if not is_shared_secret(value) else SHARED_SECRET_SERVICE
    sanitised_task_prefix = get_sanitised_kubernetes_name(task_prefix)

    secret = get_secret_name_from_ref(value)
    sanitised_secret = get_sanitised_kubernetes_name(secret)

    return V1EnvVar(
        name=key,
        value_from=V1EnvVarSource(
            secret_key_ref=V1SecretKeySelector(
                name=f"{namespace}-secret-{sanitised_task_prefix}-{sanitised_secret}",
                key=secret,
                optional=False,
            )
        ),
    )
```

would let us get rid of the duplication here :)

Reply:
i was actually asking specifically about `name` arg, docs say it's optional

Reply:
This is mostly just copied from [paasta](https://github.com/Yelp/paasta/blob/master/paasta_tools/kubernetes_tools.py#L909-L947).

I believe this name is the name of the actual V1Secret resource we want to get the value_from, since `key` is strictly the name defined within the secret data of that V1Secret resource, so we do need to keep it here (I wonder what a `SecretKeySelector` would pull from without a name).

Names of the secrets will be determined by https://github.com/Yelp/paasta/blob/master/paasta_tools/kubernetes/bin/paasta_secrets_sync.py

Reply:
ok, in that case i think this is problematic. task_processing being a library should make minimal amount of assumptions as to how entities in kubernetes (or anywhere else) are named, dependencies like that should be inverted, users of task_processing client should pass in either a function that will construct the correct name from relevant inputs or pass in a mapping where the name can be looked up.

the way this code works means that if we change how paasta generates configs (absolutely reasonable thing to do, as those are entities created and managed by paasta) will break users of task_processing. coordinating rollout of such changes will become a nightmare.

cc @jfongatyelp @nemacysts

Reply:
a followup thought - why do we even have to deal with paasta's secrets system in task_processing?

(1) ysoa-configs -> (2) paasta pre-processing -> (3) tronfig -> (4) task-processing

i think secret references should be prepared in step (2) just like we replace service/instance with correct docker image and stuff